cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/connector/resource/AwsResourceVolumeConnector.java

```diff
@@ -0,0 +1,90 @@
+import software.amazon.awssdk.services.ec2.model.DescribeVolumesRequest;
+import software.amazon.awssdk.services.ec2.model.DescribeVolumesResponse;
+import software.amazon.awssdk.services.ec2.model.DetachVolumeRequest;
+import software.amazon.awssdk.services.ec2.model.VolumeState;
+
+@Service
+public class AwsResourceVolumeConnector implements ResourceVolumeConnector {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsResourceVolumeConnector.class);
+
+    private static final int MAX_READ_COUNT = 15;
+
+    private static final int SLEEP_INTERVAL = 2;
+
+    @Override
+    public void detachVolumes(AuthenticatedContext authenticatedContext, List<CloudResource> cloudResources) throws Exception {
+        Map<String, List<Volume>> instanceVolumeIdsMap = cloudResources.stream().filter(res -> null != res.getInstanceId())
+                .collect(toMap(CloudResource::getInstanceId, resource -> resource.getParameter(CloudResource.ATTRIBUTES,
+                        VolumeSetAttributes.class).getVolumes()));
+        AmazonEc2Client amazonEC2Client = getEc2Client(authenticatedContext);
+        for (Entry<String, List<Volume>> entry: instanceVolumeIdsMap.entrySet()) {
+            entry.getValue().forEach(volume -> {
+                if (VolumeState.IN_USE.equals(getVolumeState(volume.getId(), amazonEC2Client))) {
+                    DetachVolumeRequest detachVolumeRequest = DetachVolumeRequest.builder().instanceId(entry.getKey()).volumeId(volume.getId())
+                            .device(volume.getDevice()).build();
+                    amazonEC2Client.detachVolume(detachVolumeRequest);
+                    Polling.waitPeriodly(SLEEP_INTERVAL, TimeUnit.SECONDS).stopIfException(true).stopAfterAttempt(MAX_READ_COUNT)
+                        .run(() -> {
+                            if (VolumeState.AVAILABLE.equals(getVolumeState(volume.getId(), amazonEC2Client))) {
+                                return AttemptResults.justFinish();
+                            }
+                            return AttemptResults.justContinue();
+                        });
+                }
+            });
+        }
+    }
+
+    @Override
+    public void deleteVolumes(AuthenticatedContext authenticatedContext, List<CloudResource> cloudResources) throws Exception {
+        Map<String, List<Volume>> instanceVolumeIdsMap = cloudResources.stream().filter(res -> null != res.getInstanceId())
+                .collect(toMap(CloudResource::getInstanceId, resource -> resource.getParameter(CloudResource.ATTRIBUTES,
+                        VolumeSetAttributes.class).getVolumes()));
+        AmazonEc2Client amazonEC2Client = getEc2Client(authenticatedContext);
+        for (Entry<String, List<Volume>> entry: instanceVolumeIdsMap.entrySet()) {
+            entry.getValue().forEach(volume -> {
+                DeleteVolumeRequest deleteVolumeRequest = DeleteVolumeRequest.builder().volumeId(volume.getId()).build();
```

Comment:
Please add debug log here with the volume id.

Reply:
done